ext/date/date_core.c

```diff
@@ -6039,7 +6040,7 @@ d_lite_rshift(VALUE self, VALUE other)
 			  &rm, &rd, &rjd, &ns))
 	    break;
 	if (--d < 1)
-	    rb_raise(rb_eArgError, "invalid date");
+	    rb_raise(eDateError, "invalid date");
```

Comment:
Maybe we can have more descriptive error messages?

Reply:
A good idea, but I think in this pull request, we should probably only focus on the exception class change.  A future pull request can add more descriptive exception messages.

Reply:
The main point is that right now, I don't even know if we can change the error message without breaking existing code. If code was matching the error message to detect date errors, if we changed "invalid date" to "incorrect timezone", it will cause some code to break.

As a compromise, maybe we should do "invalid date: incorrect timezone". It probably can cause issues if users used direct string comparison, but if they used `/invalid date/` maybe it's okay.

The main point I wanted to draw attention to is the fact that we should be careful not to introduce the same issue again, i.e. we have a new exception, but it's not providing enough detail for users to know exactly what is going wrong. If we know specific issues, e.g. invalid timezone, maybe we should have `Date::InvalidTimezoneError < Date::Error`.

I really like this change, and I hate to make it more complicated, but I just think if we don't consider it carefully we'll basically end up where we started.

Reply:
Any code that is trying to parse exception messages is brittle, and I don't think we should let that stop us from improving the messages.  The most recent commit changed not only exception messages, but exception types as well (a2f4b665f8ffdce83d42042676dd04cffdc45390).

Following the approach you are recommending to its logical conclusion, all exceptions with different error messages should have different exception classes. :) Now, I know you are not proposing going that far.  So the question becomes, do we think enough users will care to differentiate invalid timezones versus other forms of invalid dates.  I don't think so.  IMO, YAGNI.  And if we are gonna need it, we can add it later.